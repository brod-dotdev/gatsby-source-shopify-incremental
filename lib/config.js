"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseConfig = void 0;
const yup_1 = require("yup");
const types_1 = require("./types");
const PLUGIN_CONFIG_SCHEMA = yup_1.object({
    myshopifyDomain: yup_1.string().required(),
    adminAccessToken: yup_1.string().required(),
    storefrontAccessToken: yup_1.string().required(),
    storefrontShopDomain: yup_1.string().optional(),
    apiVersion: yup_1.mixed()
        .oneOf([types_1.ApiVersion.Apr2020, types_1.ApiVersion.Jul2020])
        .default(types_1.ApiVersion.Apr2020),
    includeCollections: yup_1.boolean().default(true),
    includePages: yup_1.boolean().default(false),
    includeBlogs: yup_1.boolean().default(false),
    shopifyPlus: yup_1.boolean().default(false),
});
function parseConfig(input) {
    var _a;
    const config = PLUGIN_CONFIG_SCHEMA.validateSync(input);
    return Object.assign(Object.assign({}, config), { storefrontShopDomain: (_a = config.storefrontShopDomain) !== null && _a !== void 0 ? _a : config.myshopifyDomain });
}
exports.parseConfig = parseConfig;
//# sourceMappingURL=config.js.map