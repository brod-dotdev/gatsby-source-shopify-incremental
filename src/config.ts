import { object, string, mixed, lazy, boolean, number, array, Schema } from "yup";
import { ApiVersion } from "./types";

/**
 * Default number of metafields to fetch
 * when no specified
 */
const DEFAULT_METAFIELD_COUNT = 100;

export interface PluginMetafieldConfig {
  /**
   * Property the metafield can be accessed on
   * Defaults to metafield{Namespace}{Key}
   */
  name: string;

  /**
   * Namespace of the Shopify metafield to include
   */
  namespace: string;

  /**
   * Key of the Shopify metafield to include
   */
  key: string;
}

/**
 * Configure the metafield fetching for a specific resource
 * - false                   => Don't fetch metafields
 * - number                  => Fetch n metafields
 * - PluginMetafieldConfig[] => Fetch these metafields
 */
export type PluginMetafieldOption = boolean | number | PluginMetafieldConfig[];

const PLUGIN_METAFIELD_CONFIG_SCHEMA = lazy<PluginMetafieldOption>((rawValue): Schema<PluginMetafieldOption> => {
  const value = rawValue as unknown;
  if (value === undefined || value === null || value === false) {
    return boolean().default(false);
  }
  if (value === true || typeof value === 'number') {
    return number().transform((value, originalValue) => { if (originalValue === true) {
      return DEFAULT_METAFIELD_COUNT;
      
    } return value }).integer();
  }
  return array(object<PluginMetafieldConfig>({
    namespace: string().required(),
    key: string().required(),
    name: string().required(),
  }))
});

/**
 * Inernal populated config
 */
export interface PluginConfig {
  /**
   * The *.myshopify.com domain of your Shopify store
   * @example my-shop.myshopify.com
   */
  myshopifyDomain: string;

  /**
   * A valid admin access token for your Shopify store
   * It must be created by the same sales channel as the
   * storefrontAccessToken
   * Required scopes read_products
   */
  adminAccessToken: string;

  /**
   * A valid storefront access token for your Shopify store
   * It must be created by the same sales channel as the
   * adminAccessToken
   */
  storefrontAccessToken: string;

  /**
   * If your storefront domain is not the default
   * *.myshopify domain
   * @example my-shop.com
   */
  storefrontShopDomain: string;

  /**
   * Shopify api version to use
   */
  apiVersion: ApiVersion;

  /**
   * Import Shopify collections
   */
  includeCollections: boolean;

  /**
   * Import Shopify pages
   */
  includePages: boolean;

  /**
   * Import Shopify blogs
   */
  includeBlogs: boolean;

  /**
   * Product metafield configuration
   */
  productMetafields: PluginMetafieldOption;

  /**
   * Indicate if you are Shopify plus
   * This will change the rate limits
   */
  shopifyPlus: boolean;
}

const PLUGIN_CONFIG_SCHEMA = object({
  myshopifyDomain: string().required(),
  adminAccessToken: string().required(),
  storefrontAccessToken: string().required(),
  storefrontShopDomain: string().optional(),
  apiVersion: mixed<ApiVersion>()
  .oneOf([ApiVersion.Apr2020, ApiVersion.Jul2020])
  .default(ApiVersion.Apr2020),
  includeCollections: boolean().default(true),
  includePages: boolean().default(false),
  includeBlogs: boolean().default(false),
  productMetafields: PLUGIN_METAFIELD_CONFIG_SCHEMA,
  shopifyPlus: boolean().default(false),
});

export function parseConfig(input: unknown): PluginConfig {
  const config = PLUGIN_CONFIG_SCHEMA.validateSync(input);

  return {
    ...config,
    storefrontShopDomain: config.storefrontShopDomain ?? config.myshopifyDomain,
  };
}
