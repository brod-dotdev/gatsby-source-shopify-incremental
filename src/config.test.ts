import { parseConfig } from './config';

const BASE_CONFIG = {
  myshopifyDomain: 'example.myshopify.com',
  adminAccessToken: 'xxxx',
  storefrontAccessToken: 'yyyy',
};

describe('config.ts', () => {
  it('passes basic config', () => {
    expect(parseConfig(BASE_CONFIG)).toMatchSnapshot('base with defaults populated');
  });

  it('correctly parses metafield config', () => {
    const mtf = [{ key: 'key', namespace: 'namespace', name: 'example' }];
    const cases = [[undefined, false], [false, false], [true, 100], [50, 50], [mtf, mtf]];

    for (let [input, output] of cases) {
      const config = parseConfig({
        ...BASE_CONFIG,
        productMetafields: input,
      });

      expect(config.productMetafields).toEqual(output)
    }
  });
});