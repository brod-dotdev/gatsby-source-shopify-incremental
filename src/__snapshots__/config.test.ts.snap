// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config.ts passes basic config: base with defaults populated 1`] = `
Object {
  "adminAccessToken": "xxxx",
  "apiVersion": "2020-04",
  "includeBlogs": false,
  "includeCollections": true,
  "includePages": false,
  "myshopifyDomain": "example.myshopify.com",
  "productMetafields": false,
  "shopifyPlus": false,
  "storefrontAccessToken": "yyyy",
  "storefrontShopDomain": "example.myshopify.com",
}
`;
